storage: cleanups after prop-eval-KV

andreimatei opened the pull request:
See individual commit messages.

cc @tschottdorf

andreimatei committed:
storage: remove dead vestiges from before prop-eval-KV

andreimatei committed:
storage: make RaftCmd.ReplicatedEvalResult non-nillable

... to suggest that it is always populated.

Also delete some vestigial code for the non-existent nil case.

andreimatei committed:
storage: clarify that various leases being passed around can't be nil

This was motivated primarily by the need to get rid of a reference to
raftCmd.BatchRequest in replica.propose (raftCmd.BatchRequest is a field
that only exists for testing now). Besides, this patch removes FUD and
takes care of some TODOs.

I've marked the lease as non-nilable in some protos, but otherwise have
continued passing leases through pointers inside a node. I've added
comments and removed unnecessary nil guards. Before writing this code,
I've sprinkled panics in the various guards that I've removed; nothing
fired in tests.

andreimatei committed:
storagebase: rename RaftCommand.BatchRequest to TestingBatchRequest

Touches #10493

andreimatei committed:
storage: bring comments on TestLeaseExtensionNotBlockedByRead up to date

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15711)
<!-- Reviewable:end -->

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -3502,7 +3502,7 @@ func (r *Replica) processRaftCommand(
 	var ts hlc.Timestamp
 	var rSpan roachpb.RSpan
-	if raftCmd.ReplicatedEvalResult != nil {
+	if idKey != "" {
```
wasn't the old code

```go
if true {
  // still here
} else if idKey != "" {
  // you deleted this
}
```

and why isn't the new code just the `// still here` part?

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -3653,30 +3639,30 @@ func (r *Replica) processRaftCommand(
 	var writeBatch *storagebase.WriteBatch
 	{
-		if raftCmd.ReplicatedEvalResult == nil && forcedErr == nil {
+		if reflect.DeepEqual(raftCmd.ReplicatedEvalResult, storagebase.ReplicatedEvalResult{}) && forcedErr == nil {
```
I really don't think there should be `reflect` in this code path. You could have `(*ReplicatedEvalResult).Empty() bool` with a corresponding test that doesn't rot when new fields are added.

tschottdorf commented on pkg/storage/storagebase/proposer_kv.proto:
```diff
@@ -172,7 +172,7 @@ message RaftCommand {
   // Proposer-evaluated KV mode.
 
-  optional ReplicatedEvalResult replicated_eval_result = 13;
+  optional ReplicatedEvalResult replicated_eval_result = 13 [(gogoproto.nullable) = false];
```
Is there a migration concern here? That is, does the previous version knows how to handle empty `ReplicatedEvalResult`? Seeing that there are various `!= nil` checks here, can one of them bite us during a rolling upgrade in which an old version receives a `ReplicatedEvalResult{}` instead of `nil` as it expects?

tschottdorf commented on pkg/storage/storagebase/state.proto:
```diff
@@ -49,7 +49,7 @@ message ReplicaState {
   // Changes of the descriptor should always go through one of the
   // (*Replica).setDesc* methods.
   roachpb.RangeDescriptor desc = 3;
-  // The latest range lease, if any.
```
optio-nit: `This field must always be populated`.

tschottdorf commented on pkg/storage/client_replica_test.go:
```diff
@@ -925,10 +925,11 @@ func TestLeaseNotUsedAfterRestart(t *testing.T) {
-// 2) Requests with the non-KV flag, such as RequestLeaseRequest, do not
-// go through the command queue.
+// 1) Lease extensions, unlike lease transfers, are not blocked by reads through
+// their ReplicatedEvanResult.BlockReads.
```
`s/Evan/Eval/`

tschottdorf commented on pkg/storage/client_replica_test.go:
```diff
@@ -925,10 +925,11 @@ func TestLeaseNotUsedAfterRestart(t *testing.T) {
-// go through the command queue.
+// 1) Lease extensions, unlike lease transfers, are not blocked by reads through
+// their ReplicatedEvanResult.BlockReads.
+// 2) Requests such as RequestLeaseRequest are not considered to touch a
```
perhaps reword to something like

> Request such as ... don't declare to touch the whole key range of the replica.

tbg reviewed (changes requested):
Thanks for cleaning all of this up. I really think we oughtn't `reflect.DeepEqual` in the hot path (or anywhere in non-test code, for sanity), but that can be fixed.
Some of your other changes around nullable fields made me nervous, but they made me nervous because of the old code. Changes to the new code in that area would make me less nervous and the extra commentary is really helpful. There are some concerns around migrations - you've probably thought about it, so hopefully they aren't serious.

I think someone else (@bdarnell?) should weigh in regarding how and when to merge this, but I for one will be happy to see it go in.

tschottdorf commented:
random addendum to this: I *think* `(ReplicatedEvalResult).IsLeaseRequest` can also go (perhaps separately, after this). And `IsFreeze`, too. And `IsConsistencyRelated`!

andreimatei commented:
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_replica_test.go, line 929 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPwE-r5-929:-KjPvzFy0Kv90sf7ixSd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9632f05ef92c411e66d662d090fb45e6db7d0d9f/pkg/storage/client_replica_test.go#L929)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`s/Evan/Eval/`
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 930 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPwE-r5-930:-KjPw5P7HqMdmSJa4X0A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9632f05ef92c411e66d662d090fb45e6db7d0d9f/pkg/storage/client_replica_test.go#L930)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

perhaps reword to something like

> Request such as ... don't declare to touch the whole key range of the replica.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 3505 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r1-3505:-KjPctHi7gwCSo3SbeqT:b-vbpgkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f59e4df839e3d2539a65bcb5d348e640e51066/pkg/storage/replica.go#L3505)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

wasn't the old code

```go
if true {
  // still here
} else if idKey != "" {
  // you deleted this
}
```

and why isn't the new code just the `// still here` part?
</blockquote></details>

:)
Because `processRaftCommand` takes "empty Raft commands", as per its documentation (and reality). Those are completely uninitialized protos. But I didn't want to complicate the proto comments with this.

---

*[pkg/storage/replica.go, line 3650 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r2-3650:-KjPsojPOnv8ixTcwq5j:b-hcov6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica.go#L3650)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I really don't think there should be `reflect` in this code path. You could have `(*ReplicatedEvalResult).Empty() bool` with a corresponding test that doesn't rot when new fields are added.
</blockquote></details>

I'll add an `Empty()` if you tell me what to test in it.
I've sprinkled `gogoproto.equal` to all the protos so we can use that generated code instead of reflection.
`proto.Equal` does not work on our protos because of we use gogoproto custom types. And it'd also be probably slower.

But the new `gogoproto.equal` generate a bunch of warnings when you do `make protobuf`, like
```
WARNING: found non-[equal] Intent with embedded equal Span
```

It's a bit unsightly. I could sprinkle some `gogoproto.equal_all` in some of our files... What do you think?

---

*[pkg/storage/storagebase/proposer_kv.proto, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPw8-r2-175:-KjPtPMWEJirXUoPt2Y_:bzceo0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/storagebase/proposer_kv.proto#L175)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Is there a migration concern here? That is, does the previous version knows how to handle empty `ReplicatedEvalResult`? Seeing that there are various `!= nil` checks here, can one of them bite us during a rolling upgrade in which an old version receives a `ReplicatedEvalResult{}` instead of `nil` as it expects?
</blockquote></details>

I don't think so... This `nullable=false` does not affect the wire format. And also, this has been filled in for a while, and we're not supposed to work with versions before that time anyway. Right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/data.pb.go, line 576 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjQ8lY4pshZmGAXC_tc:-KjQ8lY4pshZmGAXC_td:bg67lkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ef4d30d5f26d152575ddc4a1b00d13d625cbcf/pkg/roachpb/data.pb.go#L576)):*
> ```Go
> 	}
> 	return true
> }
> ```

I hope `Span.Equal` isn't used for anything performance sensitive. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 13 of 20 files at r7, 10 of 15 files at r8, 5 of 5 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 3650 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r2-3650:-KjSjzYyZ8-Dk-jLmsY5:b-gui7mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica.go#L3650)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'll add an `Empty()` if you tell me what to test in it.
I've sprinkled `gogoproto.equal` to all the protos so we can use that generated code instead of reflection.
`proto.Equal` does not work on our protos because of we use gogoproto custom types. And it'd also be probably slower.

But the new `gogoproto.equal` generate a bunch of warnings when you do `make protobuf`, like
```
WARNING: found non-[equal] Intent with embedded equal Span
```

It's a bit unsightly. I could sprinkle some `gogoproto.equal_all` in some of our files... What do you think?
</blockquote></details>

Hmm, that generated code looks unwieldy. What I was suggesting was hand-coding the `Empty` method. And that implies a test that doesn't rot when a field gets added to the proto (which as it is wouldn't force anyone to update `Empty`). So the test would be:

```go
func TestReplicatedEvalResultEmpty(t *testing.T) {
   nonEmpty := // make one that's completely nonempty
   for _, field := range fieldsOfStruct() {
      var empty ReplicatedEvalResult
      reflectionSetField(empty, nonEmpty, field) // empty.field = nonempty.field
      if empty.Empty() {
         t.Fatal()
      }
   }
}
```

That's obviously work, but there's already another location that would've needed it:

```go
	if !reflect.DeepEqual(rResult, storagebase.ReplicatedEvalResult{}) {
		log.Fatalf(ctx, "unhandled field in ReplicatedEvalResult: %s", pretty.Diff(rResult, storagebase.ReplicatedEvalResult{}))
	}
```

Of course that code has landed, so by extension, I shouldn't be berating you for yet another `DeepEqual`, though this one sits more squarely in the hot path.

Let's do the following: you simply guard the assertion:

```
if rResult.Timestamp.IsZero() && reflect.DeepEqual(...) && ... {
}
```

and add a comment that this is done to avoid the `DeepEqual` almost always.
Perhaps give the other code snippet I posted above the same treatment. I think that's a generally applicable strategy to deal with more "complicated" structs for which you can't directly compare to zero.

---

*[pkg/storage/replica.go, line 3523 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjSifA4sIUYg6NDPfbK:-KjSifA4sIUYg6NDPfbL:bs0hot9) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ef4d30d5f26d152575ddc4a1b00d13d625cbcf/pkg/storage/replica.go#L3523)):*
> ```Go
> 	// proposer_lease field in the proto.
> 	verifyLease := func() error {
> 		// Handle the case of pre-epoch-based-leases command.
> ```

This line should go, right?

---

*[pkg/storage/storagebase/proposer_kv.proto, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPw8-r2-175:-KjSu8FRHFoGQegO-b0R:b-du5ikj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/storagebase/proposer_kv.proto#L175)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think so... This `nullable=false` does not affect the wire format. And also, this has been filled in for a while, and we're not supposed to work with versions before that time anyway. Right?
</blockquote></details>

I think it does affect the wire format. `nil` doesn't get put on the wire, `ReplicatedEvalResult{}` does. So the new version, when talking to an old version, would send a `ReplicatedEvalResult{}` instead of a `nil`, which would change the . At least to the best of my knowledge.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 3650 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r2-3650:-KjTcPq739NVypdtK3gW:b3th441) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica.go#L3650)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Hmm, that generated code looks unwieldy. What I was suggesting was hand-coding the `Empty` method. And that implies a test that doesn't rot when a field gets added to the proto (which as it is wouldn't force anyone to update `Empty`). So the test would be:

```go
func TestReplicatedEvalResultEmpty(t *testing.T) {
   nonEmpty := // make one that's completely nonempty
   for _, field := range fieldsOfStruct() {
      var empty ReplicatedEvalResult
      reflectionSetField(empty, nonEmpty, field) // empty.field = nonempty.field
      if empty.Empty() {
         t.Fatal()
      }
   }
}
```

That's obviously work, but there's already another location that would've needed it:

```go
	if !reflect.DeepEqual(rResult, storagebase.ReplicatedEvalResult{}) {
		log.Fatalf(ctx, "unhandled field in ReplicatedEvalResult: %s", pretty.Diff(rResult, storagebase.ReplicatedEvalResult{}))
	}
```

Of course that code has landed, so by extension, I shouldn't be berating you for yet another `DeepEqual`, though this one sits more squarely in the hot path.

Let's do the following: you simply guard the assertion:

```
if rResult.Timestamp.IsZero() && reflect.DeepEqual(...) && ... {
}
```

and add a comment that this is done to avoid the `DeepEqual` almost always.
Perhaps give the other code snippet I posted above the same treatment. I think that's a generally applicable strategy to deal with more "complicated" structs for which you can't directly compare to zero.
</blockquote></details>

What about the generated code don't you like? It has the sort of short-circuits you suggest.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPw8-r2-175:-KjTRb5T48rlN077MeZ5:bmxfvl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/storagebase/proposer_kv.proto#L175)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think it does affect the wire format. `nil` doesn't get put on the wire, `ReplicatedEvalResult{}` does. So the new version, when talking to an old version, would send a `ReplicatedEvalResult{}` instead of a `nil`, which would change the . At least to the best of my knowledge.
</blockquote></details>

Yes, you're right. I just tested it; I was confused between `nullable=false` and `required`.

In any case, I'm curious what @bdarnell knows about this. I think all the `nil` checks that I've removed are safe. They're either about checking `ReplicatedEvalResults` produced by the same node (which should be populated), or, inside Raft, the one that I've replaced with an `.Equal(ReplicatedEvalResult{})`  we're discussing that's guarding against "really old" versions.
The current version already always populates this field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 3650 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r2-3650:-KjTpyOFGf1Lx-_eqL5X:bv57erd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica.go#L3650)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

What about the generated code don't you like? It has the sort of short-circuits you suggest.
</blockquote></details>

I don't like that it generates warnings, also the generated `Equal` always takes an `interface{}`, and it checks protobuf-semantic-equality (not that it would really make a difference here, but still). It can't guard as efficiently as picking a nearly-always-populated field manually as it doesn't know what fields those are. I think explicit guarding and `DeepEqual` is the simplest and most transparent option to check all the boxes.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPw8-r2-175:-KjTpmhPeb2JzRj5ht5p:b-7pvr9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/storagebase/proposer_kv.proto#L175)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Yes, you're right. I just tested it; I was confused between `nullable=false` and `required`.

In any case, I'm curious what @bdarnell knows about this. I think all the `nil` checks that I've removed are safe. They're either about checking `ReplicatedEvalResults` produced by the same node (which should be populated), or, inside Raft, the one that I've replaced with an `.Equal(ReplicatedEvalResult{})`  we're discussing that's guarding against "really old" versions.
The current version already always populates this field.
</blockquote></details>

Thanks, color me convinced.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'll start fixing #15385 on top of this (so please don't add extra changes/movements in this PR). Mod adding the explicit guard before the `DeepEqual` + a comment LCTM (looks concomitant to me)!

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 3650 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r2-3650:-Kj_xKKldICvb5AwWaeT:b-l1zcgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica.go#L3650)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I don't like that it generates warnings, also the generated `Equal` always takes an `interface{}`, and it checks protobuf-semantic-equality (not that it would really make a difference here, but still). It can't guard as efficiently as picking a nearly-always-populated field manually as it doesn't know what fields those are. I think explicit guarding and `DeepEqual` is the simplest and most transparent option to check all the boxes.
</blockquote></details>

I don't really like generating a bunch of code (which is probably slower than the copy we deliberately avoid in replica_proposal.go) just to power this assertion. Since this PR isn't going to land in 1.0, we should probably just remove the assertion completely (we won't need to worry about people upgrading directly from pre-release betas to 1.1). 

---

*[pkg/storage/replica.go, line 3523 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjSifA4sIUYg6NDPfbK:-Kja-2_ThwWtFIzaGaqS:bdv3vy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ef4d30d5f26d152575ddc4a1b00d13d625cbcf/pkg/storage/replica.go#L3523)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This line should go, right?
</blockquote></details>

Yeah. And now that `verifyLease` is a single expression, maybe it should just be inlined into its one call site.

---

*[pkg/storage/replica_proposal.go, line 778 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-Kj_v7wxwMQev98Ytrep:-Kj_v7wxwMQev98Ytreq:b9qu07a) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ef4d30d5f26d152575ddc4a1b00d13d625cbcf/pkg/storage/replica_proposal.go#L778)):*
> ```Go
> }
> 
> // rResult cannot be nil; it's passed by pointer to save a copy.
> ```

Making the ReplicatedEvalResult field non-nullable implies additional copies (sometimes). And we copy it below to pass to `handleReplicatedEvalResult` below. It seems strange to be concerned about copies here but not at higher or lower levels. 

---

*[pkg/storage/storagebase/proposer_kv.proto, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa4zOUiVS8PUPw8-r2-175:-Kj_uLqB47Vv7du8Nyj2:bjjep1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/storagebase/proposer_kv.proto#L175)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Thanks, color me convinced.
</blockquote></details>

It should be safe to remove the nil checks - we no longer support upgrading from pre-prop-eval-kv versions, so we should never see a nil/empty ReplicatedEvalResult (except for the empty raft commands which we guard against with the idKey check). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-:-Kja0J1a6Krsian2x9Vg:bb6ywmx)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 3523 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjSifA4sIUYg6NDPfbK:-KjbWiMEmtU_6T4iQW_z:b-ct3llu) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ef4d30d5f26d152575ddc4a1b00d13d625cbcf/pkg/storage/replica.go#L3523)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah. And now that `verifyLease` is a single expression, maybe it should just be inlined into its one call site.
</blockquote></details>

I'm doing that in my follow-up, so please leave it alone here, @andreimatei.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 3650 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjKsMa3y9oTLLwOilfD-r2-3650:-KjcrTfe-q1ZvLR5eYcR:b-m4j7o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica.go#L3650)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't really like generating a bunch of code (which is probably slower than the copy we deliberately avoid in replica_proposal.go) just to power this assertion. Since this PR isn't going to land in 1.0, we should probably just remove the assertion completely (we won't need to worry about people upgrading directly from pre-release betas to 1.1). 
</blockquote></details>

removed the assertion

---

*[pkg/storage/replica_proposal.go, line 778 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-Kj_v7wxwMQev98Ytrep:-Kjcrt5CEKlTvYjI3scG:b-nlcib7) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ef4d30d5f26d152575ddc4a1b00d13d625cbcf/pkg/storage/replica_proposal.go#L778)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Making the ReplicatedEvalResult field non-nullable implies additional copies (sometimes). And we copy it below to pass to `handleReplicatedEvalResult` below. It seems strange to be concerned about copies here but not at higher or lower levels. 
</blockquote></details>

ok, passed by value here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711)*
<!-- Sent from Reviewable.io -->

tamird commented:
This is a nice cleanup, thanks for doing it. I've left a bunch of comments, but I'll address them myself; consider them TODOs for me!

---

Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 6 of 15 files at r3, 1 of 4 files at r12, 10 of 15 files at r13, 5 of 5 files at r14, 2 of 2 files at r15.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/api.go, line 150 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdPAPdiKx-eRt7Dnxi:-KjdPAPdiKx-eRt7Dnxj:b-1oeguq) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/roachpb/api.go#L150)):*
> ```Go
> type leaseRequestor interface {
> 	// The returned Lease is never nil; returned by pointer to save a copy.
> 	prevLease() *Lease
> ```

This is inconsistent with the discussion elsewhere; why not return by value here?

---

*[pkg/storage/replica.go, line 849 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdPfq_cvJYSispFDU8:-KjdPfqaYrF-e6RLnhRN:b9yec4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/storage/replica.go#L849)):*
> ```Go
> // request initiated by this replica is in progress.
> //
> // The current lease is never nil.
> ```

This is inconsistent with discussion elsewhere; we should return the current lease by value.

---

*[pkg/storage/replica.go, line 2399 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdPkSqKYfUBh9UUaM1:-KjdPkSqKYfUBh9UUaM2:b-wgzidn) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/storage/replica.go#L2399)):*
> ```Go
> // it to the pending map.
> //
> // lease cannot be nil.
> ```

This is inconsistent with discussion elsewhere; we should pass the lease by value.

"lease" is also not the argument name here.

---

*[pkg/storage/replica.go, line 2447 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdPwZV5XewghhRcop0:-KjdPwZV5XewghhRcop1:b-uqf7ry) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/storage/replica.go#L2447)):*
> ```Go
> //   which case the other returned values are zero.
> //
> // lease cannot be nil.
> ```

This is inconsistent with discussion elsewhere; we should pass the lease by value.

---

*[pkg/storage/replica.go, line 3733 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdQ3SwNQLi5b4VIWLn:-KjdQ3SwNQLi5b4VIWLo:b8dngi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/storage/replica.go#L3733)):*
> ```Go
> 	var split *storagebase.Split
> 	var merge *storagebase.Merge
> 	split = raftCmd.ReplicatedEvalResult.Split
> ```

These should be `:=` now.

---

*[pkg/storage/replica_proposal.go, line 785 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjO5FCOTXmNFpFrwQJJ:-KjO55NDNVs6r45SU6_t:b-iq5uvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7769930d22e8dce6dd2709f02046bcd3eae4e799/pkg/storage/replica_proposal.go#L785)):*
> ```Go
> 	shouldAssert := false
> 	if rResult == nil {
> 		log.Fatalf(ctx, "nil ReplicatedEvalResult. lResult: %v", lResult)
> ```

this should just be `:=` now; there's no point to the `:= false` above anymore.

---

*[pkg/storage/replica_range_lease.go, line 336 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdQJY1u9ZjIPGt2Te8:-KjdQJY1u9ZjIPGt2Te9:b-uqf7ry) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/storage/replica_range_lease.go#L336)):*
> ```Go
> // * the client fails to read their own write.
> //
> // lease cannot be nil.
> ```

This is inconsistent with discussion elsewhere; we should pass the lease by value.

---

*[pkg/storage/storagebase/state.proto, line 52 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-KjdNFtXC4vffYCzJ-VT:-KjdNFtYYmWjpdER5ayp:b8zxich) ([raw file](https://github.com/cockroachdb/cockroach/blob/27dd36526f3b1ba703988591299de7a5edd5d151/pkg/storage/storagebase/state.proto#L52)):*
> ```protobuf
>   // (*Replica).setDesc* methods.
>   roachpb.RangeDescriptor desc = 3;
>   // The latest range lease. This field must always be populated.
> ```

I think I missed some discussion here (and reviewable isn't showing me a previously-resolved discussion in this file); how come this isn't nullable=false?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15711#-:-KjO5MKLgifs0Gtko4Vu:bpa31kb)*
<!-- Sent from Reviewable.io -->
